### PR TITLE
Font Renderer

### PR DESCRIPTION
**Pull request by 123-WILL**

### CODE REVIEW

File: src/engine/rendering/gui/fontrenderer/FontRenderer.cpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+			vertices.push_back(UiMesh::Vertex({ xpos, ypos }, { color.x, color.y, color.z, color.w }, { ch.TopLeft.x, ch.BottomRight.y }));
+
+			indices.push_back(vertices.size() - 4);
+			indices.push_back(vertices.size() - 2);
+			indices.push_back(vertices.size() - 1);
+
+			indices.push_back(vertices.size() - 2);
+			indices.push_back(vertices.size() - 3);
+			indices.push_back(vertices.size() - 1);
+
```
Comment (kevinsadi) [inline src/engine/rendering/gui/fontrenderer/FontRenderer.cpp:72]: Maybe a comment describing the high level task here would be helpful.

It looks like you're rendering out a box for each character and then rendering that TGATexture on top of it later. Does that sound right?

File: src/engine/rendering/gui/fontrenderer/TGATexture.cpp [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+		else
+		{
+			std::cerr << "Unsupported TGA format: " << bitsPerPixel << " bits per pixel" << std::endl;
+			return;
+		}
+
+		// Read pixel data
+		m_data = std::vector<unsigned char>(m_width * m_height * (bitsPerPixel / 8));
+		file.read(reinterpret_cast<char*>(m_data.data()), m_data.size());
+		file.close();
```
Comment (kevinsadi) [inline src/engine/rendering/gui/fontrenderer/TGATexture.cpp:58]: epic memory safety

File: src/engine/rendering/gui/fontrenderer/TGATexture.cpp [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+		file.read(reinterpret_cast<char*>(m_data.data()), m_data.size());
+	}
+
+	void TGATexture::LoadTexture()
+	{
+		glActiveTexture(GL_TEXTURE0);
+		
+		glGenTextures(1, &m_GlID);
+		glBindTexture(GL_TEXTURE_2D, m_GlID);
+
```
Comment (kevinsadi) [inline src/engine/rendering/gui/fontrenderer/TGATexture.cpp:145]: maybe we want to delete this code in a subsequent PR

**Review by kevinsadi — APPROVED** [review]
Great stuff man! This is going to be a giant help, with this we could probably get the timer in right now.

File: src/engine/rendering/gui/fontrenderer/FontRenderer.cpp (hunk 1 above)
Comment (123-WILL) [reply src/engine/rendering/gui/fontrenderer/FontRenderer.cpp:72]: Yeah pretty much, the UiMesh::Vertex struct has a uv built in and when it gets drawn it uses the selected font sheet texture that is stored in the font renderer class. It puts all the characters of the string you are rendering into boxes with screen and Tex coords and puts them all into one vertex/index buffer to limit draw calls

Comment (kevinsadi) [reply src/engine/rendering/gui/fontrenderer/FontRenderer.cpp:72]: Ah that checks out, so that's what `UiMesh::SetupBuffers()` was doing. Bet!

**Pull request merged**